cni/ipam/ipam_test.go

```diff
@@ -27,70 +30,191 @@ var ipamQueryResponse = "" +
 	"	</Interface>" +
 	"</Interfaces>"
 
-var localAsId string
-var poolId1 string
-var address1 string
+func TestIpam(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Ipam Suite")
+}
 
-// Wraps the test run with plugin setup and teardown.
-func TestMain(m *testing.M) {
-	var config common.PluginConfig
+// Handles queries from IPAM source.
+func handleIpamQuery(w http.ResponseWriter, r *http.Request) {
+	w.Write([]byte(ipamQueryResponse))
+}
 
-	// Create a fake local agent to handle requests from IPAM plugin.
-	u, _ := url.Parse("tcp://" + ipamQueryUrl)
-	testAgent, err := common.NewListener(u)
+func parseResult(stdinData []byte) (*cniTypesCurr.Result, error) {
+	result := &cniTypesCurr.Result{}
+	err := json.Unmarshal(stdinData, result)
 	if err != nil {
-		fmt.Printf("Failed to create agent, err:%v.\n", err)
-		return
+		return nil, err
 	}
-	testAgent.AddHandler("/", handleIpamQuery)
+	return result, nil
+}
 
-	err = testAgent.Start(make(chan error, 1))
-	if err != nil {
-		fmt.Printf("Failed to start agent, err:%v.\n", err)
-		return
-	}
+func getStdinData(cniversion, ifname, subnet, ipAddress string) []byte {
+	stdinData := fmt.Sprintf(
+`{
+			"cniversion": "%s",
+			"master": "%s",
+			"ipam": {
+				"type": "internal",
+				"subnet": "%s",
+				"ipAddress": "%s"
+			}
+		}`, cniversion, ifname, subnet, ipAddress)
+	return []byte(stdinData)
+}
 
-	// Create the plugin.
-	plugin, err = NewPlugin("ipamtest", &config)
-	if err != nil {
-		fmt.Printf("Failed to create IPAM plugin, err:%v.\n", err)
-		return
-	}
+func getIfName() (string, error) {
```

Comment:
why do we need this ? What this function expected to return? I think this is for pod interface name right?

Reply:
Yes I copy this code from network.go, for getting the interface name of the test environment.

Reply:
In cni/network/network.go 145, it will detect the in used interface name. For network.go get the same interface name as the test code, I added this function...

Reply:
the problem is if there are multiple interfaces, if it returns lo adapter or docker0 bridge as 1st interface, then this will break. for simplicity you can keep it as eth0

Reply:
Thank @DancingLinks for your clarification. We do use a real 
> the problem is if there are multiple interfaces, if it returns lo adapter or docker0 bridge as 1st interface, then this will break. for simplicity you can keep it as eth0

After re-reading the code, @DancingLinks and I found that we do not have an interface check, so both lo and docker0 will still work. But in practise, `lo` and `docker0` will not be picked for mac address mismatch.

Also, I found that interface name is not specified for the Kubernetes case[1]. How about we first cover the path without an interface name when invoking the IPAM plugin.
If we have to select a real interface name, how about removing `lo` and `docker0` from the interface list and pick one randomly, in case we may have interface names other than `eth0`, in different OS releases? This seems to be more real.

[1]https://github.com/kubernetes/kubernetes/blob/581d3e26c97ade852b54bc7f102f1a4d16b74437/pkg/kubelet/dockershim/network/cni/cni.go#L353